compositor_pipeline/src/pipeline/input/mp4.rs

```diff
@@ -0,0 +1,301 @@
+    HttpError(#[from] reqwest::Error),
+
+    #[error("Mp4 reader error.")]
+    Mp4ReaderError(#[from] mp4::Error),
+
+    #[error("No suitable track in the mp4 file")]
+    NoTrack,
+}
+
+pub struct Mp4 {
+    pub input_id: InputId,
+    video_thread: Option<std::thread::JoinHandle<()>>,
+    delete_file: bool,
+    source_file: PathBuf,
+    stop_video_reader: Arc<AtomicBool>,
+}
+
+impl Mp4 {
+    pub fn new(
+        options: Mp4Options,
+        download_dir: &Path,
+    ) -> Result<(Self, ChunksReceiver), Mp4Error> {
+        let (input_path, delete_file) = match options.source {
+            Source::Url(url) => {
+                let file_response = reqwest::blocking::get(url)?;
+                let mut file_response = file_response.error_for_status()?;
+
+                let mut path = download_dir.to_owned();
+                path.push(format!(
+                    "live-compositor-user-file-{}.mp4",
+                    rand::random::<u64>()
+                ));
+
+                let mut file = std::fs::File::create(&path)?;
+
+                std::io::copy(&mut file_response, &mut file)?;
+
+                (path, true)
+            }
+            Source::File(path) => (path, false),
+        };
+
+        let input_file = std::fs::File::open(input_path.clone())?;
+        let size = input_file.metadata()?.size();
+        let reader = mp4::Mp4Reader::read_header(input_file, size)?;
+        let stop_video_reader = Arc::new(AtomicBool::new(false));
+
+        let (video_thread, video_receiver) =
+            spawn_video_reader(reader, options.input_id.clone(), stop_video_reader.clone())?;
+
+        Ok((
+            Self {
+                input_id: options.input_id,
+                video_thread: Some(video_thread),
+                delete_file,
+                source_file: input_path,
+                stop_video_reader,
+            },
+            ChunksReceiver {
+                video: Some(video_receiver),
+                audio: None,
+            },
+        ))
+    }
+}
+
+impl Drop for Mp4 {
+    fn drop(&mut self) {
+        if let Some(thread) = self.video_thread.take() {
+            self.stop_video_reader
+                .store(true, std::sync::atomic::Ordering::Relaxed);
+            thread.join().unwrap();
+        }
+
+        if self.delete_file {
+            std::fs::remove_file(&self.source_file).unwrap();
+        }
+    }
+}
+
+fn spawn_video_reader(
+    reader: Mp4Reader<File>,
+    input_id: InputId,
+    stop_thread: Arc<AtomicBool>,
+) -> Result<(JoinHandle<()>, Receiver<EncodedChunk>), Mp4Error> {
+    let (&track_id, track, avc1) = reader
+        .tracks()
+        .iter()
+        .find_map(|(id, track)| {
+            let track_type = match track.track_type() {
+                Ok(t) => t,
+                Err(_) => return None,
+            };
+
+            let media_type = match track.media_type() {
+                Ok(t) => t,
+                Err(_) => return None,
+            };
```

Comment:
```suggestion
            let Ok(track_type) = track.track_type() else {
                return None,
            };

            let Ok(media_type) = track.media_type() else {
                return None,
            };
```
With and_then it would be even shorter, but I assume you did not use it intentionally.

Reply:
actually I just figured out I can just do this:
```rs
let track_type = track.track_type().ok()?;
```